### PR TITLE
Feature/acd 4314 fix time issue with timestamp type event

### PR DESCRIPTION
**Pull request by siddharthkhatsuriya-crest**
Changes made in the PR:
-> Add three test for indextime testing (key_fields, time, linebreaker).
-> Make HEC ingestors work with time argument.
-> change the key_fields strategy of only storing string data to be ingested in event.

### CODE REVIEW

File: pytest_splunk_addon/standard_lib/index_tests/test_templates.py [hunk 1]
```diff
@@ -58,36 +50,89 @@ def test_index_time_extractions(
...
+            {i: list(
+                set(result_fields)) for i in result_fields} == {i: list(
+                    set(fields_to_check)) for i in fields_to_check}
+        )
 
-                record_property("what_we_got_for_" + key, result[key])
+    @pytest.mark.splunk_indextime
+    def test_indextime_time(
+        self,
+        splunk_search_util,
```
Comment (jayjoshi-crest) [inline pytest_splunk_addon/standard_lib/index_tests/test_templates.py:76]: We have to add `splunk_ingest_data` to make the test case independent.

File: pytest_splunk_addon/standard_lib/index_tests/test_templates.py [hunk 2]
```diff
@@ -58,36 +50,89 @@ def test_index_time_extractions(
...
+                time_diff = int(key_time[0]) - int(event_time)
+            assert (
+                time_diff < MAX_TIME_DIFFERENCE
+            )
+
+    # Testing line breaker
+    @pytest.mark.splunk_indextime
+    def test_indextime_line_breaker(
+        self,
+        splunk_search_util,
```
Comment (jayjoshi-crest) [inline pytest_splunk_addon/standard_lib/index_tests/test_templates.py:121]: We have to add splunk_ingest_data to make the test case independent.

File: pytest_splunk_addon/standard_lib/index_tests/test_templates.py [hunk 3]
```diff
@@ -58,36 +50,89 @@ def test_index_time_extractions(
...
+        # only replaced once but the value is assigned to n events
+        # Example syslog: all the headers are only tokenized once hence
+        #   key_fields = {'host': ['dummy_host']}
+        #   result_dict = {'host': ['dummy_host']*n}
+
+        assert (
+            {i: list(
+                set(result_fields)) for i in result_fields} == {i: list(
+                    set(fields_to_check)) for i in fields_to_check}
+        )
```
Comment (jayjoshi-crest) [inline pytest_splunk_addon/standard_lib/index_tests/test_templates.py:71]: Need to add failure message.

File: pytest_splunk_addon/standard_lib/index_tests/test_templates.py [hunk 4]
```diff
@@ -58,36 +50,89 @@ def test_index_time_extractions(
...
+        event_count_query = "search sourcetype={} (host=host_{}* OR host={}*)".format(
+            splunk_indextime_line_breaker.get("sourcetype"),
+            splunk_indextime_line_breaker.get('host'),
+            splunk_indextime_line_breaker.get('host'),
         )
-        record_property("count_query", event_count_query)
+
         assert splunk_search_util.checkQueryCount(
             event_count_query, expected_events_count, retries=0, interval=0
         ), ("We should get exactly " + str(expected_events_count) + " result")
```
Comment (jayjoshi-crest) [inline pytest_splunk_addon/standard_lib/index_tests/test_templates.py:138]: In the failure message, we should add `Found {}` as well.

File: pytest_splunk_addon/standard_lib/index_tests/test_templates.py [hunk 5]
```diff
@@ -58,36 +50,89 @@ def test_index_time_extractions(
...
+
+        key_time = splunk_indextime_time["tokenized_event"].time_values
+        for index, event_time in enumerate(result_fields["e_time"]):
+            if splunk_indextime_time["tokenized_event"].metadata.get("timestamp_type") in ('event'):
+                time_diff = int(key_time[index]) - int(event_time)
+            else:
+                time_diff = int(key_time[0]) - int(event_time)
+            assert (
+                time_diff < MAX_TIME_DIFFERENCE
+            )
```
Comment (jayjoshi-crest) [inline pytest_splunk_addon/standard_lib/index_tests/test_templates.py:115]: Need to add failure message.

File: pytest_splunk_addon/standard_lib/index_tests/test_templates.py (hunk 4 above)
Comment (siddharthkhatsuriya-crest) [reply pytest_splunk_addon/standard_lib/index_tests/test_templates.py:138]: `checkQueryCount` returns boolean other ways to achieve is to fetch list of all the results compare length and show results is assertion fails, that maybe a lot of events in some cases. Anyhow we already provide the query the user can check the results in the splunk.

File: pytest_splunk_addon/standard_lib/app_test_generator.py [hunk 6]
```diff
@@ -70,17 +70,39 @@ def generate_tests(self, fixture):
             yield from self.dedup_tests(
                 self.cim_test_generator.generate_tests(fixture), fixture
             )
-        elif fixture.startswith("splunk_indextime_fields"):
+        elif fixture.startswith("splunk_indextime"):
             # TODO: What should be the id of the test case?
             # Sourcetype + Host + Key field + _count
             addon_path = self.pytest_config.getoption("splunk_app")
             sample_generator = SampleGenerator(addon_path)
```
Comment (jayjoshi-crest) [inline pytest_splunk_addon/standard_lib/app_test_generator.py:77]: Please move this logic to indextime_test_generator.  
The practice followed in searchtime is that we only call `generate_tests` method of a specific test generator from here. 
The partition was created to save complexity in a single class/method.

File: pytest_splunk_addon/standard_lib/index_tests/test_templates.py (hunk 2 above)
Comment (jayjoshi-crest) [reply pytest_splunk_addon/standard_lib/index_tests/test_templates.py:121]: Need to add `@pytest.mark.first` marker as well.

File: pytest_splunk_addon/standard_lib/index_tests/test_templates.py (hunk 1 above)
Comment (jayjoshi-crest) [reply pytest_splunk_addon/standard_lib/index_tests/test_templates.py:76]: Need to add `@pytest.mark.first` marker as well.

File: pytest_splunk_addon/standard_lib/index_tests/test_templates.py (hunk 3 above)
Comment (siddharthkhatsuriya-crest) [reply pytest_splunk_addon/standard_lib/index_tests/test_templates.py:71]: On dictionary mismatch assertion error shows what values were mismatched, do we need specific message here?

**Pull request closed without merging**